build failure due to undefined symbol with gcc 7.3.1

release-external-32 is failing to build on Fedora 27 with gcc 7.3.1:
http://dynamorio.org/CDash/buildSummary.php?buildid=35315
```
CMake Error at /work/dr/nightly/src/core/CMake_readelf.cmake:94 (message):
  *** Error:
  /.../build_release-external-32/lib32/release/libdynamorio.so
  has undefined symbol: GLOBAL DEFAULT UND __divdi3
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -580,6 +580,18 @@ if (UNIX)
 
   check_if_linker_is_gnu_gold(LINKER_IS_GNU_GOLD)
 
+  # FIXME i#2949: static 32-bit release-build linking with gcc 7.3.1 fails when
+  # static C++ clients like drmemtrace or drmemtrace_raw2trace are linked in.
+  # For now we disable those builds.
+  if (UNIX AND NOT X64 AND NOT DEBUG AND NOT CLANG AND
+      "${CMAKE_C_COMPILER_VERSION}" VERSION_GREATER "7.3")
+    message(STATUS "gcc 7.3+ non-debug 32-bit detected: disabling static C++ client "
+      "tests to work around i#2949")
+    set(DISABLE_FOR_BUG_2949 ON)
+  else ()
+    set(DISABLE_FOR_BUG_2949 OFF)
+  endif ()
+
 else (UNIX)
 
   if (NOT ${COMPILER_BASE_NAME} STREQUAL "cl")

--- a/clients/drcachesim/CMakeLists.txt
+++ b/clients/drcachesim/CMakeLists.txt
@@ -373,7 +373,6 @@ endif ()
 # We build larger executables here.  All tests are added in suite/tests/ except unit tests.
 # Be sure to give the targets qualified test names ("tool.drcache*...").
 
-
 if (BUILD_TESTS)
   add_executable(tool.drcachesim.unit_tests tests/drcachesim_unit_tests.cpp)
   if (ZLIB_FOUND)
@@ -388,7 +387,8 @@ if (BUILD_TESTS)
            COMMAND tool.drcachesim.unit_tests)
   # FIXME i#2007: fails to link on A64
   # XXX i#1997: dynamorio_static is not supported on Mac yet
-  if (NOT AARCH64 AND NOT APPLE)
+  # FIXME i#2949: gcc 7.3 fails to link certain configs
+  if (NOT AARCH64 AND NOT APPLE AND NOT DISABLE_FOR_BUG_2949)
     add_executable(tool.drcacheoff.burst_static tests/burst_static.cpp)
     configure_DynamoRIO_static(tool.drcacheoff.burst_static)
     use_DynamoRIO_static_client(tool.drcacheoff.burst_static drmemtrace_static)

--- a/core/unix/os.c
+++ b/core/unix/os.c
@@ -10383,6 +10383,48 @@ __umoddi3(uint64 dividend, uint64 divisor)
     uint64_divmod(dividend, divisor, &remainder);
     return (uint64) remainder;
 }
+
+/* Same thing for signed. */
+static int64
+int64_divmod(int64 dividend, int64 divisor64, int *remainder)
+{
+    union {
+        int64 v64;
+        struct {
+            int lo;
+            int hi;
+        };
+    } res;
+    int upper;
+    int divisor = (int) divisor64;
+
+    /* Our uses don't use large divisors. */
+    ASSERT(divisor64 <= INT_MAX && divisor64 >= INT_MIN && "divisor too large for int");
+
+    /* Divide out the high bits first. */
+    res.v64 = dividend;
+    upper = res.hi;
+    res.hi = upper / divisor;
+    upper %= divisor;
+
+    /* Like above but with the signed div instruction, which does a signed divide
+     * on edx:eax by r/m32 => quotient in eax, remainder in edx.
+     */
+    asm ("idivl %2" : "=a" (res.lo), "=d" (*remainder) :
+         "rm" (divisor), "0" (res.lo), "1" (upper));
+    return res.v64;
+}
+
+/* Match libgcc's prototype. */
+int64
+__divdi3(int64 dividend, int64 divisor)
+{
+    int remainder;
+    return int64_divmod(dividend, divisor, &remainder);
+}
+
+/* __moddi3 is coming from third_party/libgcc for x86 as well as arm. */
+
 #elif defined (ARM)
 /* i#1566: for ARM, __aeabi versions are used instead of udivdi3 and umoddi3.
  * We link with __aeabi routines from libgcc via third_party/libgcc.

--- a/suite/tests/CMakeLists.txt
+++ b/suite/tests/CMakeLists.txt
@@ -2623,7 +2623,8 @@ if (CLIENT_INTERFACE)
       # FIXME i#2007: fails to link on A64
       # XXX i#1551: startstop API is NYI on ARM
       # XXX i#1997: dynamorio_static is not supported on Mac yet
-      if (NOT AARCH64 AND NOT ARM AND NOT APPLE)
+      # FIXME i#2949: gcc 7.3 fails to link certain configs
+      if (NOT AARCH64 AND NOT ARM AND NOT APPLE AND NOT DISABLE_FOR_BUG_2949)
         torunonly_drcacheoff(burst_static tool.drcacheoff.burst_static "" "" "")
         set(tool.drcacheoff.burst_static_nodr ON)
 

